File: Medium/zigzag_conversion.py
class Solution(object):
    def convert(self, s, numRows):
        """
        :type s: str
        :type numRows: int
        :rtype: str
        """
        if len(s) == 0 or len(s) > 1000 or numRows <= 1 or numRows > 1000:
            return s
        
        collectZigzag = ["" for _ in range(numRows)]
        currRow = 0
        isStraight = True
        
        for _, c in enumerate(s):
            if currRow == 0:
                isStraight = True
            elif currRow == numRows - 1:
                isStraight = False
            
            collectZigzag[currRow] += c
                
            if isStraight:
                currRow += 1
            else:
                currRow -= 1
                

        return "".join(collectZigzag)
        
        
# print(Solution().convert("PAYPALISHIRING", 3))
print(Solution().convert("AB", 1))
        

File: Medium/zigzag_conversion.go
package main

import (
	"fmt"
	"strings"
)

func convert(s string, numRows int) string {
	if len(s) <= 1 || len(s) > 1000 || numRows <= 1 || numRows > 1000 {
		return s
	}

	var collectZigzag []string
	currRow := 0
	isStraight := true

	for i := 0; i < len(s); i++ {
		if i < numRows {
			collectZigzag = append(collectZigzag, "")
		}

		if currRow == 0 {
			isStraight = true
		} else if currRow == numRows-1 {
			isStraight = false
		}

		collectZigzag[currRow] += string(s[i])

		if isStraight {
			currRow++
		} else {
			currRow--
		}
	}

	return strings.Join(collectZigzag, "")
}

func run_convert() {
	fmt.Println(convert("PAYPALISHIRING", 3))
}
